projects/centre_console/src/race_switch.c

```diff
@@ -0,0 +1,146 @@
+}
+
+static RaceState s_event_lookup[] = { [RACE_SWITCH_EVENT_OFF] = RACE_STATE_OFF,
+                                      [RACE_SWITCH_EVENT_ON] = RACE_STATE_ON };
+
+// Triggered when the fsm switches to the normal mode
+// 5V regulator is enabled
+static void prv_state_race_off_output(Fsm *fsm, const Event *e, void *context) {
+  RaceSwitchFsmStorage *storage = (RaceSwitchFsmStorage *)context;
+  voltage_regulator_set_enabled(&storage->voltage_storage, true);
+  storage->current_state = RACE_STATE_OFF;
+}
+
+// Triggered when the fsm switches to the race mode
+// 5V regulator is disabled
+static void prv_state_race_on_output(Fsm *fsm, const Event *e, void *context) {
+  RaceSwitchFsmStorage *storage = (RaceSwitchFsmStorage *)context;
+  voltage_regulator_set_enabled(&storage->voltage_storage, false);
+  storage->current_state = RACE_STATE_ON;
+}
+
+bool race_switch_fsm_process_event(RaceSwitchFsmStorage *storage, Event *e) {
+  return fsm_process_event(&storage->race_switch_fsm, e);
+}
+
+static void prv_gpio_interrupt_handler(const GpioAddress *address, void *context) {
+  RaceSwitchFsmStorage *storage = (RaceSwitchFsmStorage *)context;
+  GpioState input_state;
+  gpio_get_state(address, &input_state);
+
+  Event e = { 0 };
+
+  // Rising edge indicates the gpio state has become high and the car should go into race mode
+  if (input_state == GPIO_STATE_HIGH) {
+    e.id = s_event_lookup[RACE_SWITCH_EVENT_ON];
+  } else {
+    e.id = s_event_lookup[RACE_SWITCH_EVENT_OFF];
+  }
+
+  race_switch_fsm_process_event(storage, &e);
+}
+
+StatusCode race_switch_fsm_init(RaceSwitchFsmStorage *storage) {
+  fsm_state_init(race_switch_off, prv_state_race_off_output);
+  fsm_state_init(race_switch_on, prv_state_race_on_output);
+
+  status_ok_or_return(gpio_init_pin(&s_race_switch_address, &s_gpio_settings));
+  status_ok_or_return(
+      gpio_init_pin(&s_voltage_regulator_enable_address, &s_voltage_enable_gpio_settings));
+  status_ok_or_return(gpio_init_pin(&s_voltage_regulator_monitor_address, &s_gpio_settings));
+
+  // Depending on whether the edge is rising or falling we can determine whether to switch into
+  // race or normal mode
+  status_ok_or_return(gpio_it_register_interrupt(&s_race_switch_address, &s_interrupt_settings,
+                                                 INTERRUPT_EDGE_RISING_FALLING,
+                                                 prv_gpio_interrupt_handler, storage));
+
+  status_ok_or_return(
+      voltage_regulator_init(&storage->voltage_storage, &s_voltage_regulator_settings));
+
+  // Determine whether the car starts in normal or race mode
+  GpioState race_switch_initial_state;
+  status_ok_or_return(gpio_get_state(&s_race_switch_address, &race_switch_initial_state));
+
+  // If in race mode (gpio state high or 1) voltage regulator must be disabled, otherwise enable
+  bool is_voltage_regulator_enabled = race_switch_initial_state == GPIO_STATE_HIGH ? false : true;
```

Comment:
Instead of checking `race_switch_initial_state == GPIO_STATE_HIGH` everywhere, I'd prefer to just have a boolean `is_race_mode` (or even a `RaceState` variable here) and check it. (Also, in general, `? false : true` can just be replaced with a boolean NOT `!`.)

As well, I'd prefer that we add a #define for the gpio state that means "race mode" (`GPIO_STATE_HIGH`), just so that in the case where we want to invert it and make `GPIO_STATE_LOW` mean "race mode", we only have to change it in one place.